code-quality/validation/README.md

```diff
@@ -0,0 +1,93 @@
+<!-- Last updated 9-25-19 -->
+
+# Input Validation 
+
+Input validation is your first line of defense when creating a secure application.  
+
+## Whitelist or Blacklist? 
+
+The first concept of good input validation is whitelisting versus blacklisting. Whitelist, or inclusive validation, defines a set of valid characters while blacklist, or exclusive validation, defines a set of invalid characters to try to remove. 
+
+Input validation using blacklisting is generally more challenging to implement so it is better practice to perform validation using whitelisting. A great way of defining a whitelist for input validation is to leverage Regular Expressions(RegEx). 
+
+## RegEx Password Validation Examples (JavaScript)
+
+- Mininum length of any 8 chars but newline: 
+```javascript
+/^.{8,}$/
+```
+
+- Check for at least one uppercase character: 
+```javascript
+/([A-Z]+)/g
+```
+
+- Check for at least one lowercase character:
+```javascript
+/([a-z]+)/g
+```
+
+- Check for at least one numeric character: 
+```javascript
+/([0-9]+)/g
+```
+
+
+- Check for at least one special character:
+```javascript
+/([!"#$%&'()*+,\-./:;<=>?@[\]^_`{|}~])+/g
+```
+
+## RegEx Email Validation Example (JavaScript)
+
+- DISCLAIMER: RegEx email validation should not be your sole source of email validation, rather it should be used to help your UI display an `invalid email` message. Whenever possible, one should include a process to confirm a given email address (e.g. send an email for the user confirm their email address).
+
+- An email is a string (a subset of ASCII characters) separated into two parts by @ symbol, a "personal_info" and a domain, that is personal_info@domain. The length of the personal_info part may be up to 64 characters long and domain name may be up to 253 characters.
+
+```javascript
+/^\w+(?:[\.-]?\w+)*@\w+(?:[\.-]?\w+)*(?:\.\w{2,24})+$/
```

Comment:
Why put a time sensitive limit on this when we know it will be out of date in a few years? Suggest `{2,}` instead to remove the upper bound, or find the max upper bound allowed by the spec, vs the max currently in use.

Reply:
This regex also does not allow for tags in email address which are *valid* such as `user+tag@mumble.dom` rather than trying to account for all the possible variants, i propose something akin to  [[anything]]@[[anything]].[[anything]] essentially, try to make this as permissive as possible, the minimum validation required to help the user from doing something stupid, all else should be handled by confirmation email.

Reply:
@DropsOfSerenity fair enough, I'll remove the upper bound for the TDL and change the regex to be more permissive.

Reply:
@codysmith287 awesome! :) This way we prevent false negatives, it can be frustrating for people with unusual email addresses or strange requirements we can't foresee (like tags) to be denied sign up for a website with a perfectly valid email address.

Reply:
@DropsOfSerenity Yeah, I agree. Thanks for the feedback! I went ahead and made those changes to email regex along with the notes below. I also added a warning right above it stating proper care should be taken to sanitize the input to prevent any SQL injection attacks.

Reply:
Stilll have some issues with this regex, prefer none at all or `/.+@.+\..+/`

Reply:
or if you like yours better, at the very least, remove the `+` after the `@`

Reply:
Also the language around sql injection is out of place being called out specifically on email, frontend validation will never stop people who want to use sql injection from the frontend, they will post the data directly where the javascript validators won't catch any of it regardless of what input you are using. That is more a backend concern, most ORM's handle it form you, I think that belongs on a backend security article and not here.

Reply:
Other than all these nitpicks, I like the rest of this PR.